### DISCUSSION TITLE
How to disable callback for add_hotkey

### DISCUSSION THREAD

**Discussion by WingTillDie** [spyoungtech/ahk#212] [Q&A] [answered]
I'm rewriting following ahk script into Python equivalent with this Python ahk library,
The following script matches current window process name,
if the current window process name is not matched, then the hotkey is not remapped
```ahk
#IfWinActive ahk_exe WindowsTerminal.exe
^n::
Send, {A}
Return
```
\
I've tried disabling callback for add_hotkey by stop_hotkeys(), but stop_hotkeys() is slow
```python
from ahk import AHK

ahk = AHK()
def my_callback():
    win = ahk.active_window
    if win.get_process_name() == 'WindowsTerminal.exe':
        ahk.send_input("a")
    else:
        ahk.stop_hotkeys() # Disable callback by stop_hotkeys()
        ahk.send_input("^n")
        ahk.start_hotkeys() # Resume callback by start_hotkeys()

ahk.add_hotkey('^n', callback=my_callback)
ahk.start_hotkeys()  # start the hotkey process thread
```
\
I suggest adding a feature that calling add_hotkey() without callback argument to disable hotkey callback, restoring to it's original behavior
It will look like this:
```python
from ahk import AHK

ahk = AHK()
def my_callback():
    win = ahk.active_window
    if win.get_process_name() == 'WindowsTerminal.exe':
        ahk.send_input("a")
    else:
        ahk.add_hotkey('^n') # Disable callback by not supplying callback argument to add_hotkey()

ahk.add_hotkey('^n', callback=my_callback)
ahk.start_hotkeys()  # start the hotkey process thread
```
\
The syntax of my suggestion is inspired by AutoIT:
```au3
#include <WinAPI.au3>

Global Const $proc_name = "WindowsTerminal.exe"

While True
	If StringInStr(_WinAPI_GetProcessName(WinGetProcess(_WinAPI_GetForegroundWindow())), $proc_name) Then
		HotKeySet("^n", "Callback")
	Else
		HotKeySet("^n")
	EndIf
	Sleep(100)
WEnd

Func Callback()
	Send("a")
EndFunc
```
\
Or are the any existing way in this Python ahk module that achieve the same feature that I want?

**Comment by spyoungtech** [comment, marked as answer]
Right now, there is no public interface for this. I am planning to add this functionality in an upcoming release.

As a workaround, you might be able to just replace the hotkey by adding the hotkey again with a no-op function under the same hotkey name. That will overwrite the existing hotkey with a new one whose callback simply does nothing, although that's not quite the same behavior as removing it.

I'll be updating progress on this in #213

**Reply by spyoungtech** [reply]
As an update on this, we've introduced public interfaces to remove/clear hotkeys and hotstrings in #215 

This is available in the release candidate `1.2.0rc1` (`pip install "ahk==1.2.0rc1"`). This will eventually be promoted and released in v1.2.0

> but stop_hotkeys() is slow

It's worth mentioning in this context that removing hotkeys/hotstrings after the hotkey subprocess is already started forces a restart of the AHK process, which will have the same "slow"ness as calling `stop_hotkeys()`. This is an implementation detail and it's possible that we can improve the performance of this in a future release.

**Reply by spyoungtech** [reply]
1.2.0 is now released with these new capabilities 🎉
